layer4/connection.go

```diff
@@ -15,21 +15,21 @@
 package layer4
 
 import (
-	"bytes"
 	"context"
-	"io"
+	"errors"
 	"net"
 	"sync"
 
 	"github.com/caddyserver/caddy/v2"
+	"go.uber.org/zap"
 )
 
 // WrapConnection wraps an underlying connection into a layer4 connection that
 // supports recording and rewinding, as well as adding context with a replacer
 // and variable table. This function is intended for use at the start of a
 // connection handler chain where the underlying connection is not yet a layer4
 // Connection value.
-func WrapConnection(underlying net.Conn, buf *bytes.Buffer) *Connection {
+func WrapConnection(underlying net.Conn, buf []byte, logger *zap.Logger) *Connection {
```

Comment:
What's wrong with `*bytes.Buffer`, exactly?

I'm also curious what the need is for a different logger on each connection?

Thanks!

Reply:
I found it a lot easier to manage and think about the offset myself instead of recreating a new Buffer each time we need to seek backwards.